Hide metadata sidebar

kenoir opened the pull request:
Hide and show metadata sidebar in search view.

![panel](https://cloud.githubusercontent.com/assets/953792/9058128/c732dc26-3a96-11e5-86d7-46c2dc5200f2.gif)

Depends: https://github.com/guardian/grid/pull/989
/cc @alastairjardine @theefer @jamesgorrie @akash1810

kenoir commented on kahuna/public/js/components/gr-panel/gr-panel.css:
```diff
@@ -3,6 +3,15 @@
+    left: 0;
+    height: calc(100vh);
+    width: 6px;
+    background-color: grey;
```
I know.

akash1810 commented:
Nice!

I personally don't like drawer handles as the toggle as the target area is usually __really__ small. The alternative is to have a button to show/hide the content - [Google Drive](https://drive.google.com/drive/my-drive) is a nice example.

I do however like the idea of collapsing. Its not of much use atm as all the operations you can do on the selection is on the (hidden) panel so would require an extra click, but once we move to the Toolbar of Action (:tm:) it would be great.

theefer commented:
> The alternative is to have a button to show/hide the content

Or handle + "hover near".

kenoir commented:
@akash1810 

> I personally don't like drawer handles as the toggle as the target area is usually really small. The alternative is to have a button to show/hide the content - Google Drive is a nice example.

I concur.

It might be worth canvassing opinion on whether it should be hidden to start with. I reckon the first action that a user is considering after selecting an item is _selecting another item_ for which having maximum screen real estate to review images is useful.

> Or handle + "hover near".

This might address the of having "another click" nicely.

akash1810 commented:
Workflow also does the "click button to open drawer" method too iirc.

akash1810 commented:
Would hover near work when the browser isn't full width?

theefer commented:
How do you mean not full width?

akash1810 commented:
When the browser is stretched to full screen then the target area is easy - just move the cursor to the far right of the screen.

If however the browser window has been resized, then the hover target area is as easy/hard to hit as an open drawer handle imo.

theefer commented:
I see, fair point. I guess the point of the hover area was that the area can be larger than the visual affordance, without taking up as much space in the UI. It's just an idea, needs to be tested.

paperboyo commented:
Wow :smile:
One thing, I don’t care much if it’s a button or handle, but I would stress the need for keyboard shortcuts (possibly, consistent in all tools, ideally to show/hide individual panels **and** to show/hide all of them at the same time). Another nice thing is when the actionable opener/closer does not move itself with the panel, so that one can show and then hide without moving the mouse (vide Composer furniture = bad; Composer Proposer Alpha = good :stuck_out_tongue_winking_eye:). That second point would make me want it on the edge, but @akash1810 is right that it may be then harder to hit when browser is not maximised. Cool! Toolbar of Action™ got me interested even more!

alastairjardine commented:
@kenoir Let's chat – I have resolved many of these issues in my head already. 

I would like to have spoken about this before it was a PR.

kenoir commented:
@alastairjardine this is a forum to discuss those issues. Have no fear, this PR does not represent releasable code or significant effort, just a suggestion.

alastairjardine commented:
A few things. 

Toggleable, hoverablable and fixable panels à la Lightroom would be nice way to go, because it nails a number of challenges in one:
* small screens can still perform contextual actions and get information without compromising # images on screen
* larger screens can have the contextual panel on display at all times
* it doesn't massively impact speed of performing actions if it appears quickly

A thin sidebar is always going to be fiddly to hit, which is why I'd suggest having a button on a secondary toolbar (not yet built), or a larger width, Proposer like bar – but that has other usability challenges if you have an appear on hover model.

I'm putting an image in to prompt a discussion. I have a clickable prototype, but I want to talk it through in person, not on Github.

**Panel — hover on button state.** Images underneath hidden.
![main search page - panel over](https://cloud.githubusercontent.com/assets/4292469/8904873/06980932-345b-11e5-8152-c71b1f7e5b6d.png)

**Panel — clicked to lock state.** Images reflow to accommodate.
![main search page - panel push](https://cloud.githubusercontent.com/assets/4292469/8904872/066f9812-345b-11e5-85df-f7aebe6cdb64.png)

kenoir commented:
@alastairjardine lets talk about this today.

A secondary toolbar would also address the concern @JonnyWeeks expressed about the new responsive top bar hiding filters while in a smaller width.

We could implement the secondary toolbar as part of this work but only use it for the metadata bar hiding/showing functionality to begin with?

JonnyWeeks commented:
hey guys

i'm not at work at the mo but .... it's awesome to see the infinite
scroller and return to search rolled out! i only noticed it yesterday but i
know the team will be absolutely stoked with that. I'll take a proper look
next time i have a good internet connection but i thought it looked great.

re the scrollbar .... As Robert says, the other thing is keeping the key
filters accessible and visible at all times. I also think grouping the
filters logically into bundles is another area that could help.

For what it's worth i like the way the guardian website's toolbar/masthead
collapses as you scroll down. Could we implement something akin to that? As
a sceondary point, personally I favour keeping it lateral because, somewhat
counterintuitively, we cannot afford for it to chew up any of the width
(because the width is precious). I know that sounds weird but if you see
how many of us work with tesselating windows (see screengrab) you'll get
what i mean.

[image: Inline images 1]

grand. great work
cheers




*Jonny Weeks*
*picture editor, **Guardian Australia*
*email: jonny.weeks@theguardian.com <jonny.weeks@theguardian.com>*
*mobile: +61(0)448018833*
*landline: +61(0)280768545*
*twitter: @WeeksJonny*
*website: **www.theguardian.com/au <http://www.theguardian.com/au>*
*address: **Level 2, 19 Foster St, Sydney, NSW 2010*

• How to supply photos to Guardian Australia <http://tinyurl.com/oqu38ap>
• How to use The Grid (for Australia staff only)
<http://tinyurl.com/nug8uor>





On 28 July 2015 at 17:23, Robert Kenny <notifications@github.com> wrote:

> @alastairjardine <https://github.com/alastairjardine> lets talk about
> this today.
>
> A secondary toolbar would also address the concern @JonnyWeeks
> <https://github.com/JonnyWeeks> expressed about the new responsive top
> bar hiding filters while in a smaller width.
>
> We could implement the secondary toolbar as part of this work but only use
> it for the metadata bar hiding/showing functionality to begin with?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/guardian/grid/pull/985#issuecomment-125481223>.
>

-- 

------------------------------
Visit theguardian.com. On your mobile and tablet, download the Guardian 
iPhone and Android apps theguardian.com/guardianapp and our tablet editions 
theguardian.com/editions.  Save up to 57% by subscribing to the Guardian 
and Observer - choose the papers you want and get full digital access. 
 Visit subscribe.theguardian.com

This e-mail and all attachments are confidential and may also be 
privileged. If you are not the named recipient, please notify the sender 
and delete the e-mail and all attachments immediately. Do not disclose the 
contents to another person. You may not use the information for any 
purpose, or store, or copy, it in any way.  Guardian News & Media Limited 
is not liable for any computer viruses or other material transmitted with 
or as part of this e-mail. You should employ virus checking software.
 
Guardian News & Media Limited is a member of Guardian Media Group plc. Registered 
Office: PO Box 68164, Kings Place, 90 York Way, London, N1P 2AP.  Registered 
in England Number 908396

alastairjardine commented:
@JonnyWeeks Some challenges to sort out with the small screen. 

Completely hear the request about being able to see what filters are set, but to resolve "filters accessible *and* visible" (my emphasis) is always going to be tricky in the scenario you posted (your image isn't coming through on github, but can see it on the email). There has to be a compromise somewhere.

I also don't think the media service scales down well in the example you showed. There are many space optimisations that could be applied to the gridded view.

JonnyWeeks commented:
Yep, totally agree that achieving accessible and visible is tricky and
compromise may be needed. My thought is that certain filters are high
priority and thus should be visible if at all poss (ie search term, free to
use, date search, uploads/recentuploads), while other stuff (ie advanced,
your uploads filter, feedback, log out etc) can afford to be dropped from
view entirely (no need even for icons) in a small browsers.

yep, there is wasted space on the smaller screens in terms of the
blank space around the thumbnails but tbh Im not too focussed on that
issue as it's not a fundamental problem for us.



*Jonny Weeks*
*picture editor, **Guardian Australia*
*email: jonny.weeks@theguardian.com
<javascript:_e(%7B%7D,'cvml','jonny.weeks@theguardian.com');>*
*mobile: +61(0)448018833*
*landline: +61(0)280768545*
*twitter: @WeeksJonny*
*website: **www.theguardian.com/au <http://www.theguardian.com/au>*
*address: **Level 2, 19 Foster St, Sydney, NSW 2010*

• How to supply photos to Guardian Australia <http://tinyurl.com/oqu38ap>
• How to use The Grid (for Australia staff only)
<http://tinyurl.com/nug8uor>





On 28 July 2015 at 18:50, Alastair Jardine <notifications@github.com
<javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:

> @JonnyWeeks <https://github.com/JonnyWeeks> Some challenges to sort out
> with the small screen.
>
> Completely hear the request about being able to see what filters are set,
> but to resolve "filters accessible *and* visible" (my emphasis) is always
> going to be tricky in the scenario you posted (your image isn't coming
> through on github, but can see it on the email). There has to be a
> compromise somewhere.
>
> I also don't think the media service scales down well in the example you
> showed. There are many space optimisations that could be applied to the
> gridded view.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/guardian/grid/pull/985#issuecomment-125508028>.
>



-- 



*Jonny Weeks*
*picture editor, **Guardian Australia*
*email: jonny.weeks@theguardian.com <jonny.weeks@theguardian.com>*
*mobile: +61(0)448018833*
*landline: +61(0)280768545*
*twitter: @WeeksJonny*
*website: **www.theguardian.com/au <http://www.theguardian.com/au>*
*address: **Level 2, 19 Foster St, Sydney, NSW 2010*

• How to supply photos to Guardian Australia <http://tinyurl.com/oqu38ap>
• How to use The Grid (for Australia staff only)
<http://tinyurl.com/nug8uor>

-- 

------------------------------
Visit theguardian.com. On your mobile and tablet, download the Guardian 
iPhone and Android apps theguardian.com/guardianapp and our tablet editions 
theguardian.com/editions.  Save up to 57% by subscribing to the Guardian 
and Observer - choose the papers you want and get full digital access. 
 Visit subscribe.theguardian.com

This e-mail and all attachments are confidential and may also be 
privileged. If you are not the named recipient, please notify the sender 
and delete the e-mail and all attachments immediately. Do not disclose the 
contents to another person. You may not use the information for any 
purpose, or store, or copy, it in any way.  Guardian News & Media Limited 
is not liable for any computer viruses or other material transmitted with 
or as part of this e-mail. You should employ virus checking software.
 
Guardian News & Media Limited is a member of Guardian Media Group plc. Registered 
Office: PO Box 68164, Kings Place, 90 York Way, London, N1P 2AP.  Registered 
in England Number 908396

kenoir commented:
@paperboyo said:

> 1. Checkboxes are z-indexed over the panel in hover state.

Needs fixing in this PR.

> 2. Hard to explain: Lock panel button's border should encompass a hot border hover area (there shouldn't be a line between them and between the panel area and Lock/Unlock panel buttons' area) and the said hot border should act together with a button on click (lock a hovered panel). In general, suddenly we have a lot of borders around buttons/toolbars/status bars.

I'll look at visually removing the border between panel and button (so it's like a tab), but I think you probably mean something else.

> 3. Changing sorting order resets selection. It shouldn't. I know it may be problematic when search exceeds a limit for images shown at any one time (the selected one would end up not being reachable). So maybe it should stay that (broken) way, until the limit is fixed/worked around.

Separate issue (needs fixing elsewhere).

> 4. Is it possible to limit scrolling with a mousewheel over the panel to only affect the panel contents? (currently when you scroll the panel all the way the main content area starts scrolling).

Separate issue (needs fixing elsewhere).

> 5. I love logic. Currently it behaves logically. If panel is unlocked it won't show on new selection (and that's its default state). That means that every time one has to open the panel after making a selection = bad. The default state should be locked. Also, against logic, but for better experience, MAYBE, even if the panel is in unlocked state, on new selection the panel should contextually open anyway (in what state, though...)?

I am deliberately clearing the locked status here as I felt it was a bit weird to do it the other way round. I think both @paperboyo and @alastairjardine disagree so i'll make the lock status persist within a session.  

> 6. Me no like sorting button placements and the fact that they explicitly state the metadata used (upload time). I'm with @akash, that the meta should be separate from the order (so that it may gain more options in the future).

I don't like the way they look at the moment either. They've ended up there as it's the least bad place until we start moving more control across (arguably they could appear in the top bar along with the search filters but we can argue about that elsewhere).

> 7. Known issue with selection disappearing from view when done further down the search results with the panel locked.

Separate issue (needs fixing elsewhere).

> 8. I'm not sure if we are using cookies or any other user prefs, but the state of the panel (lock/unlock) should persist between searches. Even betterer - between sessions. This is super-interesting in general what should persist between searches, between refreshes and between sessions in a super-web-app like yours! Just like history button.

It is interesting. Also we have a preferences API for editorial-tools :) 

> 9. Keyboard shortcuts, keyboard shortcuts! We should talk to ESD, so that we know what legacy/not used shortcuts can be freed from Fastscripts/System Prefs on Editorial machines, before thinking about a set of tools-wide shortcuts and, after that, about Grid-specific ones (ESD: Luke Hoyland + UX team).

Separate issue (needs _improving_ elsewhere).

So **todo:**

- [ ] Fix z-index issue of selection chrome over panel.
- [ ] Persist lock status within session.
- [ ] Look at removing button/panel border.

kenoir commented:
Addressed issues above:

> Fix z-index issue of selection chrome over panel.
![zindex](https://cloud.githubusercontent.com/assets/953792/9065251/758a5f52-3ac8-11e5-9191-ea559c19033f.gif)
:boom: 
> Persist lock status within session.
![locks](https://cloud.githubusercontent.com/assets/953792/9065265/84b2fc3c-3ac8-11e5-80f9-d5c1aa6e3576.gif)
:boom: :boom: 
>  Look at removing button/panel border.
![bleed](https://cloud.githubusercontent.com/assets/953792/9065283/93980a6c-3ac8-11e5-8fcc-a74034d4db39.gif)
:boom: :boom: :boom:

kenoir commented:
@akash1810 @tsop14 @theefer @jamesgorrie this defo needs a code review too.

akash1810 commented on kahuna/public/js/services/panel.js:
```diff
@@ -0,0 +1,139 @@
+            this.name = options.name;
+            this.initVisible = options.initVisible || false;
+            this.initAvailable = options.initAvailable || false;
+            this.initLocked = options.initLocked || false;
```
why do we need to assign `this.initX` rather than `this.visible = options.initVisible || false`?

akash1810 commented on kahuna/public/js/services/panel.js:
```diff
@@ -0,0 +1,139 @@
+        setAvailable,
+        setUnavailable,
+        setVisible,
+        setInvisible
```
minor - why not `show`/`hide`

paperboyo commented:
> I'll look at visually removing the border between panel and button (so it's like a tab), but I think you probably mean something else.

Thanks! Border between the button and the panel contents is one thing, although a minor one - I think it looks better without the border and is more consistent with the interactions around them.

But I also tried (and failed) to say that the button and the vertical-border-that-opens-the-panel are one. So, they should act as one: I would expect clicking on the border to realise exactly the same actions as clicking on the button (show/lock/unlock) and hover state on the border to trigger a hover state on the button and vice versa. No?

> Also we have a preferences API for editorial-tools :)

That's great! And what the cog icon's for.

kenoir commented:
@paperboyo think I get what you mean now. But it's a bit fiddly to get right :( i've removed the vertical border in the hidden state for now to stop it looking weird.

paperboyo commented:
> i've removed the vertical border in the hidden state for now

:-( I like the bar. It makes travels to open the panel shorter the further the mouse pointer is from the button. Looking forward for its big comeback ;-) No biggie, obviously!

kenoir commented:
Any objections to merge?

akash1810 commented:
@kenoir blocked on https://github.com/guardian/grid/pull/989 no?

theefer commented:
I haven't played with it yet, but so long as other people are happy. Did you it past a few power users for early feedback? Eg Joanna, Jo, etc

kenoir commented:
@theefer just shown them both. No objections to releasing it as is.

kenoir commented:
@akash1810 think we can go without the polished styling on those buttons. When #908 is released they'll get prettier.

akash1810 commented on kahuna/public/js/components/gr-panel/gr-panel.css:
```diff
@@ -5,16 +5,22 @@
 .panel--side {
     width: 284px; /* width of gr-top-bar-actions on search page */
-    height: calc(100vh - 50px);
+    height: calc(100vh - 85px);
```
can we add a comment around where `85px` comes from?

theefer commented on kahuna/public/js/components/gr-panel/gr-panel.js:
```diff
@@ -44,10 +48,28 @@ grPanel.controller('GrPanel', [
+        ctrl.isVisible = panelService.isVisible(panelName);
+
+        $rootScope.$on(
+            'ui:panels:' + panelName + ':updated',
```
FYI in general I'm not keen on dynamically generated keys for things like that because it's hard/impossible to find where things are used (if static, can just use grep). Not a biggie, but if it's always the same here, might as well just make the string static in the source?

theefer commented:
Does this branch include the order by feature? Could we roll that out first separately for ease of rollback etc?

kenoir commented:
@theefer it does. 

> Could we roll that out first separately for ease of rollback etc?

We could squish this branch before merge to make life easier if we choose to roll back, (but that might make it harder for @akash1810 to merge master).

If #908 is going out soon then we might as well wait.

kenoir commented:
After talking to @theefer.

**TODO:**

- [ ] Move number of items selected to right hand side of results bar
- [ ] Add clear selection to results bar
- [ ] Remove top clear button & number selected from gr-panel

theefer commented on kahuna/public/js/services/panel.js:
```diff
@@ -0,0 +1,136 @@
+        }
+    }
+
+    var panels = [];
```
`var` is for old people. Cool kids use `let`.

theefer replied:
Or `const`, depending on the use.

kenoir commented:
@theefer thumbs?

jamesgorrie commented on kahuna/public/js/components/gr-panel/gr-panel.html:
```diff
@@ -45,7 +43,7 @@ <h1 class="panel__title">{{ctrl.selectedImages.size}} image<span ng:if="ctrl.sel
 
                 <h2 class="image-info__title">
-                    {{ctrl.usageCategory}}
+                    {{ctrl.usageCategory || "No category specified"}}
```
Conflict with #1035

kenoir replied:
Updated to match language in that branch. Conflict should be easy to resolve.

jamesgorrie commented on kahuna/public/js/components/gr-panel/gr-panel.js:
```diff
@@ -63,6 +78,12 @@ grPanel.controller('GrPanel', [
             ctrl.rawMetadata = newMetadata;
             ctrl.images = Array.from(ctrl.selectedImages);
 
+            if (ctrl.images.length > 0) {
```
I still wonder if we should show an empty panel :bikini::house:

jamesgorrie commented:
Seems quite complex - but nice feature :

jamesgorrie commented:
:+1:

theefer commented:
Great :+1:
